File: leet_code/trees/generate_next_pointers.js
const connectNext = root => {
    if (!root) return;

    let queue = [root];

    while (queue.length > 0) {
        const len = queue.length,
              newQueue = [];
        for (let i = 0; i < len; i++) {
            const cur = queue.shift();
            queue.length === 0 ? cur.next = null : cur.next = queue[0];
            if (cur.left) newQueue.push(cur.left);
            if (cur.right) newQueue.push(cur.right);
        }
        queue = queue.concat(newQueue);
    }
};

File: leet_code/trees/generate_next_pointers.rb
def connect_next(root)
  return if (!root)

  queue = [root]

  until queue.empty?
    len = queue.length
    new_queue = []
    (0...len).each do |n|
      cur = queue.shift
      queue.empty? ? cur.next = nil : cur.next = queue[0]
      new_queue << cur.left if cur.left
      new_queue << cur.right if cur.right
    end
    queue += new_queue
  end
end
